ledger-core/v2/virtual/integration/convlog/logger.go

```diff
@@ -0,0 +1,194 @@
+	}
+	if lastIndex := strings.LastIndex(fullName, "-"); lastIndex >= 0 {
+		fullName = fullName[:lastIndex]
+	}
+
+	return fullName
+}
+
+func (v conveyorStepLogger) prepareStepName(sd *smachine.StepDeclaration) {
+	if !sd.IsNameless() {
+		return
+	}
+	sd.Name = getStepName(sd.Transition)
+}
+
+func (v conveyorStepLogger) LogUpdate(data smachine.StepLoggerData, upd smachine.StepLoggerUpdateData) {
+	special := ""
+
+	switch data.EventType {
+	case smachine.StepLoggerUpdate:
+	case smachine.StepLoggerMigrate:
+		special = "migrate "
+	default:
+		panic("illegal value")
+	}
+
+	v.prepareStepName(&data.CurrentStep)
+	v.prepareStepName(&upd.NextStep)
+
+	detached := ""
+	if data.Flags&smachine.StepLoggerDetached != 0 {
+		detached = "(detached)"
+	}
+
+	durations := ""
+	if upd.InactivityNano > 0 || upd.ActivityNano > 0 {
+		durations = fmt.Sprintf(" timing=%s/%s", upd.InactivityNano, upd.ActivityNano)
+	}
+
+	if data.Error == nil {
+		fmt.Printf("[LOG] %s[%3d]: %03d @ %03d: %s%s%s%s current=%v next=%v payload=%T tracer=%v\n", data.StepNo.MachineID(), data.CycleNo,
+			data.StepNo.SlotID(), data.StepNo.StepNo(),
+			special, upd.UpdateType, detached, durations,
+			data.CurrentStep.GetStepName(), upd.NextStep.GetStepName(), v.sm, v.tracer)
+		return
+	}
+
+	errSpecial := ""
+	switch data.Flags & smachine.StepLoggerErrorMask {
+	case smachine.StepLoggerUpdateErrorMuted:
+		errSpecial = "muted "
+	case smachine.StepLoggerUpdateErrorRecovered:
+		errSpecial = "recovered "
+	case smachine.StepLoggerUpdateErrorRecoveryDenied:
+		errSpecial = "recover-denied "
+	}
+
+	fmt.Printf("[ERR] %s[%3d]: %03d @ %03d: %s%s%s%s current=%v next=%v payload=%T tracer=%v%s\n", data.StepNo.MachineID(), data.CycleNo,
+		data.StepNo.SlotID(), data.StepNo.StepNo(),
+		special, errSpecial, upd.UpdateType, detached, data.CurrentStep.GetStepName(), upd.NextStep.GetStepName(), v.sm, v.tracer,
+		formatErrorStack(data.Error))
+}
+
+func (v conveyorStepLogger) LogInternal(data smachine.StepLoggerData, updateType string) {
+	v.prepareStepName(&data.CurrentStep)
+
+	fmt.Printf("[ERR] %s[%3d]: %03d @ %03d: internal %s current=%v payload=%T tracer=%v%s\n", data.StepNo.MachineID(), data.CycleNo,
+		data.StepNo.SlotID(), data.StepNo.StepNo(),
+		updateType, data.CurrentStep.GetStepName(), v.sm, v.tracer, formatErrorStack(data.Error))
+}
+
+func (v conveyorStepLogger) LogEvent(data smachine.StepLoggerData, customEvent interface{}, fields []logfmt.LogFieldMarshaller) {
+	special := ""
+
+	v.prepareStepName(&data.CurrentStep)
+
+	switch data.EventType {
+	case smachine.StepLoggerTrace:
+		special = "TRC"
+	case smachine.StepLoggerActiveTrace:
+		special = "TRA"
+	case smachine.StepLoggerWarn:
+		special = "WRN"
+	case smachine.StepLoggerError:
+		special = "ERR"
+	case smachine.StepLoggerFatal:
+		special = "FTL"
+	default:
+		fmt.Printf("[U%d] %s[%3d]: %03d @ %03d: current=%v payload=%T tracer=%v custom=%v\n", data.EventType, data.StepNo.MachineID(), data.CycleNo,
```

Comment:
this should be logger from call context

Reply:
this logger is purely for local debug and made to be **readable**, not for logs